src/nunavut/lang/cpp/templates/_fields_as_union.j2

```diff
@@ -0,0 +1,13 @@
+    {{ composite_type.tag_field_type | declaration }} tag;
+    union {
+{%- for field in composite_type.fields -%}
+{%- if field is not padding %}
+        {{ field.data_type | declaration }} {{ field.name | id }};
+{%- endif -%}
+{%- endfor %}
+    };
+{%- for field in composite_type.fields -%}
+{%- if field is not padding %}
+    bool {{ ('is_' + field.name)  | id }}() { return (tag == {{ loop.index0 }}); }
+{%- endif -%}
+{%- endfor %}
```

Comment:
If we chose to go this way, then we should hide the inner union and the tag from the user completely by making them private. Exposing the aliased memory is unsafe and MISRA linters will be complaining about that in the user code. We need getters (const and not) that return a pointer to the member if the tag is correct and nullptr otherwise (they will also obviate `is_...()`). A setter per field is also needed to allow switching the tag value.

The above is undesirable though. Ideally, we should avoid reinventing this little wheel completely and just use `std::variant` as all normal people do, but it will cause issues with C++14/11. To avoid said issues, we could use `std::variant` if the detected version is C++17 or newer, otherwise, fallback to [`bpstd::variant`](https://github.com/bitwizeshift/BackportCpp/blob/master/include/bpstd/variant.hpp) or some other alternative. Handling raw unions without the typesafe infrastructure that is readily available for `std::variant` will be hard, error-prone, and require much manual boilerplating.

Reply:
I'm going to comment this out so we can get this right in the future but not expose the current implementation to the public (without warning). New change incoming...